### PR TITLE
Adding deaggregation logic

### PR DESCRIPTION
**Pull request by BizzlePop34**

### CODE REVIEW

File: lib/polling-consumer.js [hunk 1]
```diff
@@ -136,7 +137,9 @@ async function pollForRecords(instance) {
       return;
     }
 
-    const records = await Promise.all(Records.map(recordsDecoder));
+    const deaggCollection = await deaggregate(Records);
+    const records = await Promise.all(deaggCollection.map(recordsDecoder));
```
Comment (jennyEckstein) [inline lib/polling-consumer.js:141]: Don't see a limit for Promise.all, maybe use p-map?

File: lib/deaggregate.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+/*!
+ * Code adpted from aws-kinesis-agg 4.0.4
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+const ProtoBuf = require('protobufjs');
+const aggJson = require('./aggregate-protobuf.json');
+
+const MAGIC_NUMBER = 'f3899ac2';
```
Comment (jennyEckstein) [inline lib/deaggregate.js:12]: I checked the documentation, and absolutely hate that they call it MAGIC_NUMBER. Can we rename the variable KPL_MAGIC_NUMBER, so its more clear where it comes from.

File: lib/deaggregate.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * @param {object} kinesisRecord - The kinesis message
+ * @param {Function} perRecordCallback - A callback invoked for each deaggregated record
+ * @param {Function} afterRecordCallback - A callback invoked after all records have been deaggregated
+ */
+const deaggregate = (kinesisRecord, perRecordCallback, afterRecordCallback) => {
+  // we receive the record data as a base64 encoded string
+  const recordBuffer = Buffer.from(kinesisRecord.Data, 'base64');
+
+  // first 4 bytes are the kpl assigned magic number
```
Comment (jennyEckstein) [inline lib/deaggregate.js:30]: instead of kpl can spell out Kinesis Producer Library (KPL)

File: lib/deaggregate.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      // emit the per-record callback with the extracted partition
+      // keys and sequence information
+      perRecordCallback(null, {
+        ApproximateArrivalTimestamp: kinesisRecord.ApproximateArrivalTimestamp,
+        Data: item.data,
+        ExplicitPartitionKey: protobufMessage.explicit_hash_key_table[item.explicit_hash_key_index],
+        PartitionKey: protobufMessage.partition_key_table[item.partition_key_index],
+        SequenceNumber: kinesisRecord.SequenceNumber,
+        SubSequenceNumber: i
+      });
```
Comment (jennyEckstein) [inline lib/deaggregate.js:50]: Let destructure kinesisRecord, so we can clean up this object a little.

File: lib/deaggregate.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    // user records vs plain Kinesis Records on the basis of the
+    // sub-sequence number
+
+    perRecordCallback(null, {
+      ApproximateArrivalTimestamp: kinesisRecord.ApproximateArrivalTimestamp,
+      Data: kinesisRecord.Data,
+      ExplicitPartitionKey: kinesisRecord.ExplicitPartitionKey,
+      PartitionKey: kinesisRecord.PartitionKey,
+      SequenceNumber: kinesisRecord.SequenceNumber
+    });
```
Comment (jennyEckstein) [inline lib/deaggregate.js:67]: destructure kinesisRecord would make it a bit cleaner.

File: lib/deaggregate.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+module.exports = (kinesisRecords) => {
+  const userRecords = [];
+  return new Promise((resolve) => {
+    // use the async deaggregation interface, and accumulate user records into
+    // the userRecords array
+
+    if (!kinesisRecords || kinesisRecords.length === 0) {
+      resolve(kinesisRecords);
+
```
Comment (jennyEckstein) [inline lib/deaggregate.js:80]: let remove this empty line

File: lib/deaggregate.test.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      partition_key_table: ['key1', 'key2'],
+      records: messages
+    };
+    const buff = builder.encode(message).finish();
+
+    const md5 = crypto.createHash('md5');
+    md5.update(buff);
+    const checksum = md5.digest();
+    content = {
+      Data: Buffer.concat([Buffer.from('f3899ac2', 'hex'), buff, checksum]),
```
Comment (jennyEckstein) [inline lib/deaggregate.test.js:42]: let put 'f3899ac2' in a constant

File: lib/deaggregate.test.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    expect(Buffer.from(result[3].Data).toString()).toEqual('message2');
+  });
+
+  test('null values are handled properly', async () => {
+    const result = await deaggregate(null);
+    expect(result).toEqual(null);
+  });
+
+  test('case where the records are not aggregated', async () => {
+    const deaggRecord = {
```
Comment (jennyEckstein) [inline lib/deaggregate.test.js:78]: Since this object is only used once, we can place it directly in deaggregate function

**Review by jennyEckstein — COMMENTED** [review]
Looks very nice, I found some minor changes

**Review by eaviles — CHANGES_REQUESTED** [review]
@BizzlePop34 changes look OK but let's make de-aggregation optional just like in the KPL/KCL. Can you please add a "useDeaggregation" (or something similar) to the public constructor?

**Pull request merged**